node/operations.js

```diff
@@ -208,6 +216,46 @@ Operations.prototype.addInReq = function addInReq(req) {
     return req;
 };
 
+Operations.prototype.hasDrained = function hasDrained() {
+    var self = this;
+
+    if (self.pending.in === 0 &&
+        self.pending.out === 0) {
+        return true;
+    } else if (self.drainExempt &&
+               self._isCollDrained(self.requests.in) &&
+               self._isCollDrained(self.requests.out)) {
+        return true;
+    }
+
+    return false;
+};
+
+Operations.prototype.checkDrained = function checkDrained() {
+    var self = this;
+
+    if (self.hasDrained()) {
+        self.drainEvent.emit(self);
+        self.drainEvent.removeAllListeners();
+    }
+};
+
+Operations.prototype._isCollDrained = function _isCollDrained(coll) {
+    var self = this;
+
+    /* jshint forin:false */
+    for (var id in coll) {
+        var op = coll[id];
+        if (!(op instanceof OperationTombstone) &&
```

Comment:
Let's check `op.drained` as well.

It's possible to add an outgoing request in the "drained" state to this operation.